### PR TITLE
feature(controller): 4 - put together the components for the upgraded AER Controller

### LINKED ISSUE #233: feat: improve seeding interface for Cycle
At the moment, the seeding of the SimpleCycle needs a lot of manual work to set up. The initial experiment, data, and theory all potentially need manual initialization.

The cycle with the StateMachine as its steering element makes this easier, but it turned out to be somewhat complicated to implement.

Is there a better way?

### PR DESCRIPTION
**Pull request by hollandjg**
## Description

Create new `Controller`s which supersede the `SimpleCycle`. 

This is part of the solution to:
- #238 
- #247 
- #248

This builds on:
- #299 
- #303
- #304

- resolves #233

Todos:
- [ ] Convert `__init__.py` into a notebook and include in documentation. Can we still use it as doctests?

### Type of change:
- New feature (non-breaking change which adds functionality)

### Features: 
- **Added a `BaseController`** (`base.py`)
   Handles shared logic (how to get the next step in the cycle, for instance) for all the controller classes
- **Added a `Cycle`**, derived from the `BaseController` (`cycle.py`)
   Has the same interface as the `SimpleCycle` but uses the `Snapshot` state from #299, the `full_cycle_planner` from #303, and the `default_online_executor_collection` from #304 
- **Added a `Controller`**, derived from the `BaseController` (`controller.py`), 
   Offers easier seeding and a history function than the `Cycle`, and the ability to choose a different `planner` function. It uses the `Hitory` state from #299, the `last_result_kind_planner` (as a default) from #303, and the `default_online_executor_collection` from #304. 
   Resolves #233
- **Added documentation** showing all the classes in use (`__init__.py`): 
  - simple execution using the `Cycle`, 
  - more complex execution and customization using the `Controller` and 
  - full customization and multiple experimentalists by using the `BaseController`

### CODE REVIEW

File: autora/controller/__init__.py [hunk 1]
```diff
@@ -0,0 +1,559 @@
...
+
+    ... or the precision (here we keep iterating while the difference between the gradients
+    of the second-last and last cycle is larger than 1x10^-3).
+    >>> _ = list(
+    ...         takewhile(
+    ...             lambda c: np.abs(c.data.theories[-1].coef_.item() -
+    ...                            c.data.theories[-2].coef_.item()) > 1e-3,
+    ...             cycle
+    ...         )
+    ...     )
```
Comment (musslick) [inline autora/controller/__init__.py:134]: This is neat!

File: autora/controller/__init__.py [hunk 2]
```diff
@@ -0,0 +1,559 @@
...
+    MONITOR: Generated new CONDITION
+    MONITOR: Generated new OBSERVATION
+    MONITOR: Generated new THEORY
+    MONITOR: Generated new CONDITION
+    MONITOR: Generated new OBSERVATION
+    MONITOR: Generated new THEORY
+
+    However, if we seed the same cycle with observations, then its first Executor will be the
+    theorist:
+    >>> controller_with_seed_observation = Controller(
```
Comment (musslick) [inline autora/controller/__init__.py:307]: Love it!

**Review by musslick — CHANGES_REQUESTED** [review]
I think this looks fantastic. Especially the ability to define your own planner and introduce contingencies about which experimentalist/theorist to use dependent on the history object. 

The examples in the init are really well documented and this makes we wonder whether we could supply a python notebook here that we could easily integrate in the documentation (or on Google Collab)? I think this would help users a lot. We could literally use the __init__.py examples as the notebook. What do you think?

**Comment by hollandjg** [comment]
> I think this looks fantastic. Especially the ability to define your own planner and introduce contingencies about which experimentalist/theorist to use dependent on the history object.
> 
> The examples in the init are really well documented and this makes we wonder whether we could supply a python notebook here that we could easily integrate in the documentation (or on Google Collab)? I think this would help users a lot. We could literally use the **init**.py examples as the notebook. What do you think?

Yes, we totally should. I'm not yet sure the best way to do that and integrate the notebook with the documentation though.

**Comment by hollandjg** [comment]
this is too big. Closing for now, will move this to a separate repo. Thanks for the feedback!

**Pull request closed without merging**